yabause/src/qt/doc/yabause.1
.TH YABAUSE 1 "January 7, 2013" "yabause-0.9.12"
.SH NAME
yabause \- Yet Another Buggy And Uncomplete Saturn Emulator
.SH SYNOPSIS
.B yabause
[ \fB\-afh\fP ] [ \fB\-b <BIOS>\fP | \fB\-nb\fP ] [ \fB\-c <CDROM>\fP | \fB\-i <ISO>\fP ] [ \fB-ns\fP ]
[ \fB\-\-autoframeskip=0|1\fP ]
[ \fB\-\-autoload=<SLOT>\fP ]
[ \fB\-\-autostart\fP ]
[ \fB\-\-binary=<BINARY>[:<ADDRESS>]\fP ]
.SH DESCRIPTION
\fBYabause\fP is a Sega Saturn emulator. \fBYabause\fP needs either a bios file, a game or a binary to run.
Games can be loaded from a real cd device or from dump files.
.SH OPTIONS
.TP
.BI \-\-autoframeskip
Enable or disable auto frame skipping / limiting.
.TP
.BI \-\-autoload
Automatically start emulation and load a savestate.
.TP
.BI \-a
.TP
.BI \-\-autostart
Automatically start emulation.
.TP
.BI \-b
.TP
.BI \-\-bios
Choose a bios file.
.TP
.BI \-c
.TP
.BI \-\-cdrom
Choose the cdrom device.
.TP
.BI \-\-binary
Use a binary file. Content of the file will be loaded to 0x06004000 and execution will start from that address.
You can provide an alternative address to load and run the binary with the \-\-binary=<BINARY>:<ADDRESS> syntax.
When using this option, emulation is automatically started and you shouldn't use it in conjunction with \-a.
This option is intended for homebrew developers wanting to test their programs in \fBYabause\fP.
.TP
.BI \-f
.TP
.BI \-\-fullscreen
Start the emulator in fullscreen.
.TP
.BI \-i
.TP
.BI \-\-iso
Choose a dump file. The dump can be either in iso or bin/cue file format.
.TP
.BI \-nb
.TP
.BI \-\-no\-bios
Use the emulated bios.
.TP
.BI \-ns
.TP
.BI \-\-no\-sound
Turns sound off. This option actually set the sound core to the dummy one.
.SH FILES
.TP
\fB<BINARY>\fR
An executable for the Saturn, usually with a BIN extension.
.TP
\fB<BIOS>\fR
A Saturn ROM BIOS image.
.TP
\fB<CDROM>\fR
A CDROM device file.
.TP
\fB<ISO>\fR
A Saturn game dump.
.SH AUTHORS
Copyright (c) 2002-2013 Yabause Team

Web:	http://yabause.org

Please don't ask for roms, bios files or any other
copyrighted stuff.
.SH COPYRIGHT
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as 
published by the Free Software Foundation; either version 2 of 
the License, or (at your option) any later version.

This program is distributed in the hope that it will be
useful,but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public
License along with this program; if not, write to the Free
Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301  USA

See the GNU General Public License details in COPYING.


yabause/ChangeLog
0.9.11 -> 0.9.12
 general:
  - Fixes to the dynamic recompiler (Ari64)
  - Added ARMv5 support to the dynarec (Ari64)
  - New OSD system (Guillaume)
  - Added "built-in" DDK to make it easier to compile on Windows (Guillaume)
 sound:
  - Improvements / Fixes in both SCSP and SCSP2 (Cwiiis)
 video:
  - Major improvements to the OpenGL renderer (Devmiyax)
  - Major improvements to the software renderer (Guillaume)
  - Some fixes to register emulation (Guillaume)
  - Improvements to line drawing functions in the software renderer (Cwiiis)
  - Fixed endianess bugs (Guillaume)
 cocoa port:
  - Added "load image" feature (BlueCrab)
  - Fixed the resize bug (BlueCrab)
 qt port:
  - Added shortcuts to toggle vdp2 layers (Benjamin Siskoo)
  - Fixed the "mute sound" feature (Guillaume)
  - It's now possible to compile the Qt port in "full software" mode (Guillaume)
  - Added an autostart option, disabled by default (Guillaume)
  - Now using a XDG compliant location for config file (Guillaume)
  - Added a debugger to the Qt port (CyberWarriorX)
  - DirectX cores can now be used in Qt port (CyberWarriorX)
  - Cheat search function (CyberWarriorX)
  - Option to show/hide menu and toolbar (Guillaume)
  - Close button in pad settings (guillaume)
0.9.10 -> 0.9.11
 general:
  - Now using CMake as the default build system.
    For now, autotools based build and "custom" build systems are still supported.
  - New Cocoa port
  - Added a dynamic recompiling SH2 core for x86 and ARM
  - New SCSP implementation
  - Major update of the software renderer from the yabause-rr team
  - Added an option to allow to execute from the cache
  - Improvements to the OpenGL renderer
 carbon port:
  - Improvements
 gtk port:
  - Added command line option to enable/disable frame skipping / limiting.
  - Added frame skipping/limiting configuration in settings.
  - Added --autoload command line option
  - Vdp2 layers can be toggled from the Vdp2 debug window
 psp port:
  - Added support for Media Engine CPU
  - Improvements to the PSP port 
 qt port:
  - Added command line support
 wii port:
  - Merged some stuff from the wiibrew fork, mostly related to SH2 emulation
 windows port:
  - Fixed the XBox controller driver
  - Fixed the "open iso then cancel bug"
0.9.9 -> 0.9.10
 scsp/68k:
  - Added code to make SCSP emulation frame-accurate (optional,
    enabled with --enable-scsp-frame-accurate configure switch).
  - Added a new 68000 emulation core.
 software video core:
  - Added line scroll emulation.
  - Improved user clipping.
  - Added some basic vertical scroll emulation, enough to get
    Sonic Jam working.
 gtk port
  - Gtk port is now compiling on Mac OS X.
  - Fixed full software screenshots.
  - Fixed store function in transfer dialog.
 windows port:
  - Added 12 player support.
  - Fixed a bug that was causing the memory transfer dialog
    to register the wrong filename after pressing "Browse".
  - Fixed bugs in Goto Address dialog.
  - Fixed a bug that was causing the vdp2 viewer dialog
    to register the wrong filename after pressing "Browse".
  - Added MD saving in SCU DSP debug dialog.
  - Added new Ram Watch dialog.
  - Added video recording feature.
  - Added move recording feature.
 general:
  - Added Lithuanian translation.
  - New sound core using OpenAL.
  - Added joystick core for Mac OS X.
  - Added a joystick core for Linux.
  - Added a PSP port.
  - Added support for loading ELF binaries.
  - Now using gettimeofday when available for better resolution.
  - Fixed save states.
0.9.8 -> 0.9.9
 opengl video core:
  - Fixed a bug that was causing some games to
    crash (albert odyssey, dragon ball, etc.)
 gtk port:
  - Automatic detection of current locale.
 qt port:
  - Added support of DESTDIR and --program-prefix
  - Automatic detection of current locale.
  - Added support for multiple players.
 windows port:
  - Fixed crash when going into settings.
  - Fixed mouse wheel usage in disassembler.
  - Rewrote as an unicode application.
  - Fixed the key configuration problem.
  - Fixed joypad support.
  - Partial fix for mouse wheel and slider problem.
  - Fixed fullscreen bug.
 general:
  - Hooks for renaming .desktop on installation.
  - .yts file are now installed.
  - Fixed parallel builds.
  - SDL peripheral core now handles all connected
    joysticks.
0.9.7 -> 0.9.8
 vdp2:
  - Fixed a bug in software renderer with rotating
    backgrounds.
 opengl video core:
  - Added gouraud shading and mesh processing.
    This is not enabled by default.
 software video core:
  - Fixed user clipping.
 gtk port:
  - Added mouse support.
  - Configuration dialog now displays key names instead
    of values. Also made it so each different configuration
    is saved. This broke compatibility with old .ini files.
 windows port:
  - Support for spaces in filenames when using CLI.
  - Added mouse support.
  - Added cheat search.
 general:
  - Added mouse emulation.
  - Added de, es, it, pt-br and sv translations
  - Support for "out of src" build.
  - Fixed compilation for non supported platforms.
    For instance this should fix compilation on dragonfly bsd.
    Fixed compilation on GNU/Hurd too.
0.9.6 -> 0.9.7
 vdp1:
  - Added clipping for line-based drawing to software renderer.
 vdp2:
  - Toggling a screen is now core independent.
  - Added per-character priority to software renderer.
 gtk port:
  - Fixed fullscreen setting and added a keep ratio one.
  - Fixed a bug in the vdp2 debugger that was causing the emu to crash.
  - Full software mode can be compiled again.
  - Fixed segfault when taking screenshots in full software mode.
  - Fixed default value for region.
  - Window position is now saved and restored when re-opening the emu.
  - Fixed a problem when changing input cores.
 qt port:
  - Improved compilation process: make (un)install now works.

yabause/src/android/src/org/yabause/android/Yabause.java
/*  Copyright 2011 Guillaume Duhamel

    This file is part of Yabause.

    Yabause is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Yabause is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Yabause; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
*/

package org.yabause.android;

import java.lang.Runnable;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.graphics.Bitmap;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MenuInflater;
import android.app.Dialog;
import android.app.AlertDialog;
import android.app.AlertDialog.Builder;
import android.content.DialogInterface;
import org.yabause.android.YabauseView;
import android.widget.ImageView;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;

class InputHandler extends Handler {
    private YabauseRunnable yr;

    public InputHandler(YabauseRunnable yr) {
        this.yr = yr;
    }

    public void handleMessage(Message msg) {
        if (msg.arg1 == 1) {
            yr.press(msg.arg2);
        } else if (msg.arg1 == 2) {
            yr.release(msg.arg2);
        }
    }
}

class YabauseRunnable implements Runnable
{
    public static native int init(Yabause yabause, Bitmap bitmap);
    public static native void deinit();
    public static native void exec();
    public static native void press(int key);
    public static native void release(int key);
    public static native int initViewport( int width, int hieght);
    public static native int drawScreen();
    public static native int lockGL();
    public static native int unlockGL();
    
    private boolean inited;
    private boolean paused;
    public InputHandler handler;

    public YabauseRunnable(Yabause yabause, Bitmap bitmap)
    {
        handler = new InputHandler(this);
        int ok = init(yabause, bitmap);
        inited = (ok == 0);
    }

    public void pause()
    {
        Log.v("Yabause", "pause... should really pause emulation now...");
        paused = true;
    }

    public void resume()
    {
        Log.v("Yabause", "resuming emulation...");
        paused = false;
        handler.post(this);
    }

    public void destroy()
    {
        Log.v("Yabause", "destroying yabause...");
        inited = false;
        deinit();
    }

    public void run()
    {
        if (inited && (! paused))
        {
            exec();
            
            handler.post(this);
        }
    }

    public boolean paused()
    {
        return paused;
    }
}

class YabauseHandler extends Handler {
    private Yabause yabause;

    public YabauseHandler(Yabause yabause) {
        this.yabause = yabause;
    }

    public void handleMessage(Message msg) {
        yabause.showDialog(msg.what, msg.getData());
    }
}

public class Yabause extends Activity implements OnTouchListener
{
    private static final String TAG = "Yabause";
    private YabauseRunnable yabauseThread;
    private YabauseHandler handler;

    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.main);

        YabauseView view = (YabauseView) findViewById(R.id.yabause_view);
        handler = new YabauseHandler(this);
        yabauseThread = new YabauseRunnable(this,null);
        view.setYabauseRunnable(yabauseThread);

        ImageView pad = (ImageView) findViewById(R.id.yabause_pad);
        pad.setOnTouchListener(this);
    }

    @Override
    public void onPause()
    {
        super.onPause();
        Log.v(TAG, "pause... should pause emulation...");
        yabauseThread.pause();
    }

    @Override
    public void onResume()
    {
        super.onResume();
        Log.v(TAG, "resume... should resume emulation...");
        yabauseThread.resume();
    }

    @Override
    public void onDestroy()
    {
        super.onDestroy();
        Log.v(TAG, "this is the end...");
        yabauseThread.destroy();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.emulation, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
        case R.id.pause:
            yabauseThread.pause();
            return true;
        case R.id.quit:
            this.finish();
            return true;
        case R.id.resume:
            yabauseThread.resume();
            return true;
        default:
            return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        if (yabauseThread.paused()) {
            menu.setGroupVisible(R.id.paused, true);
            menu.setGroupVisible(R.id.running, false);
        } else {
            menu.setGroupVisible(R.id.paused, false);
            menu.setGroupVisible(R.id.running, true);
        }
        return true;
    }

    @Override
    public Dialog onCreateDialog(int id, Bundle args) {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage(args.getString("message"))
            .setCancelable(false)
            .setNegativeButton("Exit", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                    Yabause.this.finish();
                }
            })
            .setPositiveButton("Ignore", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                    dialog.cancel();
                }
            });
        AlertDialog alert = builder.create();
        return alert;
    }

    public boolean onTouch(View v, MotionEvent event) {
        int action = event.getActionMasked();
        float x = event.getX();
        float y = event.getY();
        int keyx = (int) ((x - 10) / 30);
        int keyy = (int) ((y - 10) / 30);
        int key = (keyx << 2) | keyy;
        int keya = 0;
        if (action == event.ACTION_DOWN) {
            keya = 1;
        } else if (action == event.ACTION_UP) {
            keya = 2;
        }

        Message message = handler.obtainMessage();
        message.arg1 = keya;
        message.arg2 = key;
        yabauseThread.handler.sendMessage(message);

        return true;
    }

    private void errorMsg(String msg) {
        Message message = handler.obtainMessage();
        Bundle bundle = new Bundle();
        bundle.putString("message", msg);
        message.setData(bundle);
        handler.sendMessage(message);
    }

    static {
        System.loadLibrary("yabause");
    }
}
